itests/kit/node_unmanaged.go

```diff
@@ -464,332 +697,427 @@ func (tm *TestUnmanagedMiner) SubmitProveCommit(
+		}
+		partitionMap[sp.Partition] = append(partitionMap[sp.Partition], sector)
+	}
 
-	chainRandomnessEpoch := di.Challenge
-	chainRandomness, err := tm.FullNode.StateGetRandomnessFromTickets(ctx, crypto.DomainSeparationTag_PoStChainCommit, chainRandomnessEpoch,
+	chainRandomnessEpoch := challengeEpoch
+	chainRandomness, err := tm.FullNode.StateGetRandomnessFromTickets(tm.ctx, crypto.DomainSeparationTag_PoStChainCommit, chainRandomnessEpoch,
 		nil, head.Key())
 	if err != nil {
-		return fmt.Errorf("Miner(%s): failed to get chain randomness for sector %d: %w", tm.ActorAddr, sectorNumber, err)
+		return fmt.Errorf("Miner(%s): failed to get chain randomness for deadline %d: %w", tm.ActorAddr, di.Index, err)
 	}
 
-	minerInfo, err := tm.FullNode.StateMinerInfo(ctx, tm.ActorAddr, head.Key())
+	minerInfo, err := tm.FullNode.StateMinerInfo(tm.ctx, tm.ActorAddr, head.Key())
 	if err != nil {
-		return fmt.Errorf("Miner(%s): failed to get miner info for sector %d: %w", tm.ActorAddr, sectorNumber, err)
+		return fmt.Errorf("Miner(%s): failed to get miner info: %w", tm.ActorAddr, err)
 	}
 
-	r, err := tm.SubmitMessage(ctx, &miner14.SubmitWindowedPoStParams{
-		ChainCommitEpoch: chainRandomnessEpoch,
-		ChainCommitRand:  chainRandomness,
-		Deadline:         sp.Deadline,
-		Partitions:       []miner14.PoStPartition{{Index: sp.Partition}},
-		Proofs:           []proof.PoStProof{{PoStProof: minerInfo.WindowPoStProofType, ProofBytes: proofBytes}},
-	}, 0, builtin.MethodsMiner.SubmitWindowedPoSt)
-	if err != nil {
-		return fmt.Errorf("Miner(%s): failed to submit window post for sector %d: %w", tm.ActorAddr, sectorNumber, err)
+	postMessages := make([]cid.Cid, 0)
+
+	submit := func(partitions []uint64, sectors []sectorInfo) error {
+		var proofBytes []byte
+		if tm.mockProofs {
+			proofBytes = []byte{0xde, 0xad, 0xbe, 0xef}
+		} else {
+			proofBytes, err = tm.generateWindowPost(sectors)
+			if err != nil {
+				return fmt.Errorf("Miner(%s): failed to generate window post for deadline %d, partitions %v: %w", tm.ActorAddr, di.Index, partitions, err)
+			}
+		}
+
+		tm.log("WindowPoST submitting %d sectors for deadline %d, partitions %v", len(sectors), di.Index, partitions)
+
+		pp := make([]miner14.PoStPartition, len(partitions))
+		for i, p := range partitions {
+			pp[i] = miner14.PoStPartition{Index: p}
+		}
+
+		// Push all our post messages to the mpool and wait for them later. The blockminer might have
+		// ceased mining while it waits to see our posts in the mpool and if we wait for the first
+		// message to land but it's waiting for a later sector then we'll deadlock.
+		mCid, err := tm.mpoolPushMessage(&miner14.SubmitWindowedPoStParams{
+			ChainCommitEpoch: chainRandomnessEpoch,
+			ChainCommitRand:  chainRandomness,
+			Deadline:         di.Index,
+			Partitions:       pp,
+			Proofs:           []proof.PoStProof{{PoStProof: minerInfo.WindowPoStProofType, ProofBytes: proofBytes}}, // can only have 1
```

Comment:
I see. So we don't need to associate sector proofs with their corresponding partition here. You can just say "here are the partitions I am proving" and "here is one single proof for all the sectors in all those partitions".

Reply:
Yes; and on top of that, when you actually submit it, you have this `Proofs` property which is an array but it can only have 1 in it: https://github.com/filecoin-project/builtin-actors/blob/54236ae89880bf4aa89b0dba6d9060c3fd2aacee/actors/miner/src/lib.rs#L510-L516

I'm not sure what the history is here, but I think you can have multiple for a `VerifyWindowPoSt` call, so you could make the most of that if you wanted to verify your own posts locally.